server/server.go

```diff
@@ -0,0 +1,214 @@
+	return func(s *Server) error {
+		s.initializers = append(s.initializers, initializerFunc)
+		return nil
+	}
+}
+
+// WithGrpcServer adds the given GRPC server to this server. There can only be one GRPC server within a given instance,
+// so multiple calls with this option will overwrite the previous ones.
+func WithGrpcServer(grpcServer *grpc.Server) Option {
+	return func(s *Server) error {
+		s.GRPCServer = grpcServer
+		return nil
+	}
+}
+
+// WithHandler registers the given http handler to this server by registering the pattern at the root of the http server
+func WithHandler(pattern string, handler http.Handler) Option {
+	return func(s *Server) error {
+		s.registrars = append(s.registrars, func(mux *http.ServeMux) error {
+			mux.Handle(pattern, handler)
+			return nil
+		})
+		return nil
+	}
+}
+
+// WithHealthChecks registers the given health checker with this server by registering its endpoints at the root of the
+// http server.
+func WithHealthChecks(checker health.Checker) Option {
+	return func(s *Server) error {
+		s.registrars = append(s.registrars, func(mux *http.ServeMux) error {
+			checker.RegisterHandler(mux)
+			return nil
+		})
+		return nil
+	}
+}
+
+// WithGateway registers the given gateway options with this server
+func WithGateway(options ...gateway.Option) Option {
+	return func(s *Server) error {
+		s.registrars = append(s.registrars, func(mux *http.ServeMux) error {
+			_, err := gateway.NewGateway(append(options, gateway.WithMux(mux))...)
+			return err
+		})
+		return nil
+	}
+}
+
+// Serve invokes all initializers then serves on the given listeners.
+//
+// If a listener is left blank, then that particular part will not be served.
+//
+// If a listener is specified for a part that doesn't have a corresponding server, then an error will be returned. This
+// can happen, for instance, whenever a gRPC listener is provided but no gRPC server was set or no option was passed
+// into NewServer.
+func (s *Server) Serve(grpcL, httpL net.Listener) error {
+	if err := s.initialize(); err != nil {
+		return err
+	}
+	errC := make(chan error)
+
+	if httpL != nil {
+		if s.HTTPServer == nil {
+			return errors.New("httpL is specified, but no HTTPServer is provided")
+		}
+		go func() { errC <- s.HTTPServer.Serve(httpL) }()
+	} else {
+		s.HTTPServer = nil
+	}
+
+	if grpcL != nil {
+		if s.GRPCServer == nil {
+			return errors.New("grpcL is specified, but no GRPCServer is provided")
+		}
+		go func() { errC <- s.GRPCServer.Serve(grpcL) }()
+	} else {
+		s.GRPCServer = nil
+	}
+
+	return <-errC
+}
+
+func (s *Server) Stop() error {
+	doneC := make(chan bool)
+	errC := make(chan error)
+	go func() {
+		defer func() { doneC <- true }()
+		if s.GRPCServer != nil {
+			s.GRPCServer.Stop()
+		}
+	}()
+	go func() {
+		defer func() { doneC <- true }()
+		if s.HTTPServer != nil {
+			if err := s.HTTPServer.Close(); err != nil {
+				errC <- err
+			}
+		}
+	}()
+
+	select {
+	case err := <-errC:
+		return err
+	case <-doneC:
```

Comment:
Disclaimer: I'm not a concurrency expert, but I don't think is quite right. Shouldn't the `select` continue until both the HTTP server _and_ the gRPC are closed? In this case, the first server that closes will trigger the `return` statement.

If this isn't the behavior you want, then I think [buffered channels](https://tour.golang.org/concurrency/3) will come in handy.

Reply:
This is what I wanted -- if either of the servers go down, then the entire thing should go down. That way you're not running a gateway without the gRPC server, or vice-versa.

Reply:
good find! 
the `http.Server` struct also has both [`Shutdown`](https://golang.org/pkg/net/http/#Server.Shutdown) and [`Close`](https://golang.org/pkg/net/http/#Server.Close) functions, so I could build out the two functions for both services